ops/lost-in-translation.js

```diff
@@ -0,0 +1,163 @@
+  const locales = filenames.map(filename => parse(filename).name);
+  return locales;
+}
+
+/**
+ * Add rows to a Google Spreadsheet that are not already added.
+ */
+async function addUniqueRowsToGoogleSheet(sheet, rowsToAdd) {
+  const alreadyAddedRows = await sheet.getRows({ limit: 1000 });
+  const uniqueRowsToAdd = rowsToAdd.filter(rowToAdd =>
+    !alreadyAddedRows.find(alreadyAddedRow => alreadyAddedRow.key === rowToAdd.key));
+  await sheet.addRows(uniqueRowsToAdd);
+  return uniqueRowsToAdd;
+}
+
+/**
+ * Step 1: Find all the (english) translation keys across all branches and PRs.
+ *
+ * If a Dutch developer has made a feature in a branch, we expect that him/her added a key
+ * to one or many locale files (usually the english locale (en.json) or to their own locale
+ * file (nl.json), or both).
+ *
+ * But they probably haven't added translations to all the other locales, because they don't
+ * speak the other languages. And this is the problem, because now we need find people to help
+ * translate the added keys to all the other locales.
+ *
+ * So what we do here is simply to find *all* the english translation keys across the entire
+ * project. That means looking at all the english translations keys in all the locale files
+ * since the start of the project across all branches and PRs. We throw them into a set that
+ * we use in the next step.
+ */
+const allLocales = retrieveAllLocales('app/locales/');
+const allEnglishTranslationKeys = new Set([]);
+for (const locale of allLocales) {
+  const filePath = `app/locales/${locale}.json`;
+  for (const commitHash of findCommitHashesForFile(filePath)) {
+    const translation = retrieveJSONForFileAtCommitHash(filePath, commitHash);
+    for (const translationKey of Object.keys(translation)) {
+      allEnglishTranslationKeys.add(normalizeTranslationKey(translationKey));
+    }
+  }
+}
+
+/**
+ * Step 2: For each locale file check if there are any missing english translation keys
+ *         across all branches and PRs and all commits/changes.
+ *
+ * If there's a missing english translation key, we know that we're most likely missing
+ * a translation for that locale. So what we need to do is to translate it, or get help
+ * to translate it.
+ * 
+ * We throw it into a dictionary where the key is the locale and the value is a set of 
+ * missing translations from english to that locale.
+ */
+const translationKeysByLocale = {};
+for (const locale of allLocales) {
+  const filePath = `app/locales/${locale}.json`;
+  for (const commitHash of findCommitHashesForFile(filePath)) {
+    const translation = retrieveJSONForFileAtCommitHash(filePath, commitHash);
+    const translationKeys = Object.keys(translation).map(key => normalizeTranslationKey(key));
+    if (translationKeysByLocale[locale] === undefined) {
+      translationKeysByLocale[locale] = new Set([]);
+    }
+    translationKeysByLocale[locale] = new Set([...translationKeysByLocale[locale], ...translationKeys]);
```

Comment:
I feel stuff like this is mentally much easier to visualize when `translationKeysByLocale ` is typed :)

Reply:
Yeah we can tweak that, wanted to get it out quickly before bed